protocols/autonat/src/v2/server/handler/dial_request.rs

```diff
@@ -0,0 +1,331 @@
+                        self.rng.clone(),
+                    ))
+                    .is_err()
+                {
+                    tracing::warn!(
+                        "failed to push inbound request handler, too many requests in flight"
+                    );
+                }
+            }
+            ConnectionEvent::ListenUpgradeError(ListenUpgradeError { error, .. }) => {
+                tracing::debug!("inbound request failed: {:?}", error);
+            }
+            _ => {}
+        }
+    }
+}
+
+enum HandleFail {
+    InternalError(usize),
+    RequestRejected,
+    DialRefused,
+    DialBack {
+        idx: usize,
+        result: Result<(), DialBackStatus>,
+    },
+}
+
+impl From<HandleFail> for DialResponse {
+    fn from(value: HandleFail) -> Self {
+        match value {
+            HandleFail::InternalError(addr_idx) => Self {
+                status: ResponseStatus::E_INTERNAL_ERROR,
+                addr_idx,
+                dial_status: DialStatus::UNUSED,
+            },
+            HandleFail::RequestRejected => Self {
+                status: ResponseStatus::E_REQUEST_REJECTED,
+                addr_idx: 0,
+                dial_status: DialStatus::UNUSED,
+            },
+            HandleFail::DialRefused => Self {
+                status: ResponseStatus::E_DIAL_REFUSED,
+                addr_idx: 0,
+                dial_status: DialStatus::UNUSED,
+            },
+            HandleFail::DialBack { idx, result } => Self {
+                status: ResponseStatus::OK,
+                addr_idx: idx,
+                dial_status: match result {
+                    Err(DialBackStatus::DialErr) => DialStatus::E_DIAL_ERROR,
+                    Err(DialBackStatus::DialBackErr) => DialStatus::E_DIAL_BACK_ERROR,
+                    Ok(()) => DialStatus::OK,
+                },
+            },
+        }
+    }
+}
+
+async fn handle_request(
+    stream: impl AsyncRead + AsyncWrite + Unpin,
+    observed_multiaddr: Multiaddr,
+    client: PeerId,
+    dial_back_cmd_sender: mpsc::Sender<DialBackCommand>,
+    rng: impl RngCore,
+) -> Event {
+    let mut coder = Coder::new(stream);
+    let mut all_addrs = Vec::new();
+    let mut tested_addr_opt = None;
+    let mut data_amount = 0;
+    let response = handle_request_internal(
+        &mut coder,
+        observed_multiaddr.clone(),
+        dial_back_cmd_sender,
+        rng,
+        &mut all_addrs,
+        &mut tested_addr_opt,
+        &mut data_amount,
+    )
+    .await
+    .unwrap_or_else(|e| e.into());
+    if tested_addr_opt.is_none() {
+        return Event {
+            all_addrs,
+            tested_addr: observed_multiaddr,
+            client,
+            data_amount,
+            result: Err(io::Error::new(
+                io::ErrorKind::Other,
+                "client is not conformint to protocol. the tested address is not the observed address",
+            )),
+        };
+    }
+    let tested_addr = tested_addr_opt.unwrap();
```

Comment:
This can be a `let-else` to avoid the unwrap.

```suggestion
    let Some(tested_addr) = tested_addr_opt else {
        return Event {
            all_addrs,
            tested_addr: observed_multiaddr,
            client,
            data_amount,
            result: Err(io::Error::new(
                io::ErrorKind::Other,
                "client is not conformint to protocol. the tested address is not the observed address",
            )),
        };
    };
```

Reply:
good point, what elegant use of syntax. clearly GitHub's syntax highlighting couldn't keep up.